dbms/src/Storages/StorageS3.cpp

```diff
@@ -0,0 +1,177 @@
+#include <Storages/StorageFactory.h>
+#include <Storages/StorageS3.h>
+
+#include <Interpreters/Context.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Parsers/ASTLiteral.h>
+
+#include <IO/ReadBufferFromS3.h>
+#include <IO/WriteBufferFromS3.h>
+
+#include <Formats/FormatFactory.h>
+
+#include <DataStreams/IBlockOutputStream.h>
+#include <DataStreams/IBlockInputStream.h>
+#include <DataStreams/AddingDefaultsBlockInputStream.h>
+
+#include <Poco/Net/HTTPRequest.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+namespace
+{
+    class StorageS3BlockInputStream : public IBlockInputStream
+    {
+    public:
+        StorageS3BlockInputStream(const Poco::URI & uri,
+            const String & format,
+            const String & name_,
+            const Block & sample_block,
+            const Context & context,
+            UInt64 max_block_size,
+            const ConnectionTimeouts & timeouts)
+            : name(name_)
+        {
+            read_buf = std::make_unique<ReadBufferFromS3>(uri, timeouts);
+
+            reader = FormatFactory::instance().getInput(format, *read_buf, sample_block, context, max_block_size);
+        }
+
+        String getName() const override
+        {
+            return name;
+        }
+
+        Block readImpl() override
+        {
+            return reader->read();
+        }
+
+        Block getHeader() const override
+        {
+            return reader->getHeader();
+        }
+
+        void readPrefixImpl() override
+        {
+            reader->readPrefix();
+        }
+
+        void readSuffixImpl() override
+        {
+            reader->readSuffix();
+        }
+
+    private:
+        String name;
+        std::unique_ptr<ReadBufferFromS3> read_buf;
+        BlockInputStreamPtr reader;
+    };
+
+    class StorageS3BlockOutputStream : public IBlockOutputStream
+    {
+    public:
+        StorageS3BlockOutputStream(const Poco::URI & uri,
+            const String & format,
+            const Block & sample_block_,
+            const Context & context,
+            const ConnectionTimeouts & timeouts)
+            : sample_block(sample_block_)
+        {
+            auto minimum_upload_part_size = context.getConfigRef().getUInt64("s3_minimum_upload_part_size", 512'000'000);
```

Comment:
Ok. But is has base-2 and base-10 mixed value.
Maybe 512 * 1024 * 1024 will be way more cool.

Reply:
I prefer it to be table level, not global setting.

Reply:
#7059